projects/wepiggy.js

```diff
@@ -2,55 +2,111 @@
+    oracle: '0x4c78015679fabe22f6e02ce8102afbf7d93794ea',
+  },
+};
 
-  async function getUnderlying(token) {
-    if (token === '0x27A94869341838D5783368a8503FdA5fbCd7987c') {
-      return '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2';//pETH => WETH
-    }
-    const contract = new web3.eth.Contract(abi.underlying, token)
-    return await contract.methods.underlying().call();
-  }
+// ask comptroller for all markets array
+async function getAllMarkets(block, chain, comptroller) {
+  const { output: markets } = await sdk.api.abi.call({
+    target: comptroller,
+    abi: abi['getAllMarkets'],
+    block,
+    chain: chain,
+  });
+  return markets;
+}
 
-  async function getCash(token) {
-    let contract = new web3.eth.Contract(abi.getCash, token);
-    let cash = await contract.methods.getCash().call();
-    let underlying = await getUnderlying(token);
-    let decimals = await utils.returnDecimals(underlying)
-    let price = (await utils.getTokenPricesFromString(underlying)).data[underlying.toLowerCase()].usd
-    cash = await new BigNumber(cash).div(10 ** decimals).times(price);
-    return cash;
+async function getUnderlyingDecimals(block, chain, token) {
+  // if (token === '0x27A94869341838D5783368a8503FdA5fbCd7987c') {
+  //   return { underlying: '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2', decimals: 18 }; //pETH => WETH
+  // }
+  if (token.toLowerCase() === '0x27A94869341838D5783368a8503FdA5fbCd7987c'.toLowerCase()) {
+    return 18; //ETH
   }
+  if (token.toLowerCase() === '0x621ce6596e0b9ccf635316bfe7fdbc80c3029bec'.toLowerCase()) {
+    return 18; //OKT
+  }
+  const { output: underlying } = await sdk.api.abi.call({
+    target: token,
+    abi: abi['underlying'],
+    block,
+    chain: chain,
+  });
+  const { output: decimals } = await sdk.api.abi.call({
+    target: underlying,
+    abi: abi['decimals'],
+    block,
+    chain: chain,
+  });
+  return decimals;
+}
 
-  async function fetch() {
-
-    let tvl = new BigNumber('0');
+async function getUnderlyingPrice(block, chain, oracle, token) {
+  const { output: underlyingPrice } = await sdk.api.abi.call({
+    target: oracle,
+    abi: abi['getUnderlyingPrice'],
+    block,
+    params: [token],
+    chain: chain,
+  });
+  return underlyingPrice;
+}
 
-    let allMarkets = await getAllMarkets();
+async function getCash(block, chain, token) {
+  const { output: cash } = await sdk.api.abi.call({
+    target: token,
+    abi: abi['getCash'],
+    block,
+    chain: chain,
+  });
+  return cash;
+}
 
-    await (
-      Promise.all(allMarkets.map(async (token) => {
-        let cash = await getCash(token)
-        tvl = tvl.plus(cash)
-      }))
-    );
+async function fetch() {
+  let tvl = new BigNumber('0');
+  await Promise.all(
+    Object.keys(contracts).map(async key => {
+      let { comptroller, oracle } = contracts[key];
+      let chain = key;
+      let block = null;
+      // const { block } = await sdk.api.util.lookupBlock(timestamp, {
```

Comment:
Did you comment this out because okex rpc is not an archive node?

Reply:
Because there are no parameters for fetch function. Tvl function has timestamp parameters but it is required to return the Balances object, which temporarily cannot solve the price problem. So we need to commented this first and use it later.

Reply:
Makes sense. Thanks for submitting this, I'll got ahead and merge!